File: packages/memory/src/queue.ts
/**
 * @fileoverview In-memory queue storage
 * @module @anyq/memory/queue
 */

import type { MessageHeaders, ProviderMetadata } from '@anyq/core';
import { generateMessageId } from '@anyq/core';

/**
 * Internal message storage format
 */
export interface StoredMessage<T = unknown> {
  id: string;
  body: T;
  key?: string;
  headers: MessageHeaders;
  timestamp: Date;
  deliveryAttempt: number;
  acknowledged: boolean;
  requeued: boolean;
  deadLettered: boolean;
}

/**
 * In-memory queue implementation
 *
 * Provides a simple FIFO queue for testing and development.
 */
export class MemoryQueue<T = unknown> {
  private messages: StoredMessage<T>[] = [];
  private processing: Map<string, StoredMessage<T>> = new Map();
  private readonly queueName: string;
  private readonly maxMessages: number;
  private readonly maxAgeMs: number;

  constructor(
    queueName: string = 'default',
    options: { maxMessages?: number; maxAgeMs?: number } = {}
  ) {
    this.queueName = queueName;
    this.maxMessages = options.maxMessages ?? 0;
    this.maxAgeMs = options.maxAgeMs ?? 0;
  }

  /**
   * Add a message to the queue
   */
  enqueue(
    body: T,
    options: { key?: string; headers?: MessageHeaders } = {}
  ): string {
    // Clean up old messages if needed
    this.cleanup();

    const id = generateMessageId();
    const message: StoredMessage<T> = {
      id,
      body,
      key: options.key,
      headers: options.headers ?? {},
      timestamp: new Date(),
      deliveryAttempt: 0,
      acknowledged: false,
      requeued: false,
      deadLettered: false,
    };

    this.messages.push(message);

    // Enforce max messages limit
    if (this.maxMessages > 0 && this.messages.length > this.maxMessages) {
      this.messages.shift();
    }

    return id;
  }

  /**
   * Get the next message from the queue
   */
  dequeue(): StoredMessage<T> | undefined {
    this.cleanup();

    const message = this.messages.shift();
    if (!message) {
      return undefined;
    }

    message.deliveryAttempt++;
    this.processing.set(message.id, message);
    return message;
  }

  /**
   * Peek at the next message without removing it
   */
  peek(): StoredMessage<T> | undefined {
    return this.messages[0];
  }

  /**
   * Get multiple messages from the queue
   */
  dequeueBatch(count: number): StoredMessage<T>[] {
    this.cleanup();

    const messages: StoredMessage<T>[] = [];
    for (let i = 0; i < count; i++) {
      const message = this.dequeue();
      if (!message) break;
      messages.push(message);
    }
    return messages;
  }

  /**
   * Acknowledge a message
   */
  ack(messageId: string): boolean {
    const message = this.processing.get(messageId);
    if (!message) {
      return false;
    }

    message.acknowledged = true;
    this.processing.delete(messageId);
    return true;
  }

  /**
   * Negative acknowledge - requeue or discard
   */
  nack(messageId: string, requeue: boolean = true): boolean {
    const message = this.processing.get(messageId);
    if (!message) {
      return false;
    }

    this.processing.delete(messageId);

    if (requeue) {
      message.requeued = true;
      this.messages.unshift(message);
    }

    return true;
  }

  /**
   * Move message to dead letter queue
   */
  deadLetter(
    messageId: string,
    dlq: MemoryQueue<unknown>,
    error?: Error
  ): boolean {
    const message = this.processing.get(messageId);
    if (!message) {
      return false;
    }

    this.processing.delete(messageId);

    // Add error info to headers
    const dlqHeaders: MessageHeaders = {
      ...message.headers,
      'x-original-queue': this.queueName,
      'x-death-reason': error?.message ?? 'max retries exceeded',
      'x-death-time': new Date().toISOString(),
      'x-delivery-attempts': String(message.deliveryAttempt),
    };

    dlq.enqueue(message.body as unknown, {
      key: message.key,
      headers: dlqHeaders,
    });

    return true;
  }

  /**
   * Get queue size
   */
  size(): number {
    return this.messages.length;
  }

  /**
   * Get number of messages being processed
   */
  processingCount(): number {
    return this.processing.size;
  }

  /**
   * Check if queue is empty
   */
  isEmpty(): boolean {
    return this.messages.length === 0;
  }

  /**
   * Clear all messages
   */
  clear(): void {
    this.messages = [];
    this.processing.clear();
  }

  /**
   * Get all messages (for debugging)
   */
  getAll(): StoredMessage<T>[] {
    return [...this.messages];
  }

  /**
   * Clean up expired messages
   */
  private cleanup(): void {
    if (this.maxAgeMs <= 0) {
      return;
    }

    const now = Date.now();
    this.messages = this.messages.filter(
      (msg) => now - msg.timestamp.getTime() < this.maxAgeMs
    );
  }

  /**
   * Create provider metadata for a message
   */
  createMetadata(): ProviderMetadata {
    return {
      provider: 'memory',
      memory: {
        queueName: this.queueName,
      },
    };
  }
}

File: go/memory/queue.go
// Package memory is an in-process queue adapter for anyq, used for testing and
// development and as the reference implementation for the retry-strategy hooks.
package memory

import (
	"sync"
	"time"

	"github.com/sns45/anyq/go/core"
)

// StoredMessage is the internal storage format for an in-memory message.
type StoredMessage struct {
	ID              string
	Body            []byte
	Key             string
	Headers         core.MessageHeaders
	Timestamp       time.Time
	DeliveryAttempt int
	Acknowledged    bool
	Requeued        bool
	DeadLettered    bool
}

// Queue is a simple FIFO in-memory queue with a processing set.
type Queue struct {
	mu          sync.Mutex
	name        string
	messages    []*StoredMessage
	processing  map[string]*StoredMessage
	maxMessages int
	maxAge      time.Duration
}

// NewQueue builds a Queue.
func NewQueue(name string, maxMessages int, maxAge time.Duration) *Queue {
	if name == "" {
		name = "default"
	}
	return &Queue{
		name:        name,
		processing:  make(map[string]*StoredMessage),
		maxMessages: maxMessages,
		maxAge:      maxAge,
	}
}

// Name returns the queue name.
func (q *Queue) Name() string { return q.name }

// Enqueue adds a message and returns its id.
func (q *Queue) Enqueue(body []byte, key string, headers core.MessageHeaders) string {
	q.mu.Lock()
	defer q.mu.Unlock()
	q.cleanup()

	if headers == nil {
		headers = core.MessageHeaders{}
	}
	id := core.GenerateMessageID()
	msg := &StoredMessage{
		ID:        id,
		Body:      body,
		Key:       key,
		Headers:   headers,
		Timestamp: time.Now(),
	}
	q.messages = append(q.messages, msg)
	if q.maxMessages > 0 && len(q.messages) > q.maxMessages {
		q.messages = q.messages[1:]
	}
	return id
}

// Dequeue removes and returns the next message, moving it into processing.
func (q *Queue) Dequeue() *StoredMessage {
	q.mu.Lock()
	defer q.mu.Unlock()
	q.cleanup()
	if len(q.messages) == 0 {
		return nil
	}
	msg := q.messages[0]
	q.messages = q.messages[1:]
	msg.DeliveryAttempt++
	q.processing[msg.ID] = msg
	return msg
}

// DequeueBatch returns up to count messages.
func (q *Queue) DequeueBatch(count int) []*StoredMessage {
	out := make([]*StoredMessage, 0, count)
	for i := 0; i < count; i++ {
		m := q.Dequeue()
		if m == nil {
			break
		}
		out = append(out, m)
	}
	return out
}

// Ack acknowledges a processing message.
func (q *Queue) Ack(id string) bool {
	q.mu.Lock()
	defer q.mu.Unlock()
	msg, ok := q.processing[id]
	if !ok {
		return false
	}
	msg.Acknowledged = true
	delete(q.processing, id)
	return true
}

// Nack returns a message to the queue (requeue) or discards it.
func (q *Queue) Nack(id string, requeue bool) bool {
	q.mu.Lock()
	defer q.mu.Unlock()
	msg, ok := q.processing[id]
	if !ok {
		return false
	}
	delete(q.processing, id)
	if requeue {
		msg.Requeued = true
		q.messages = append([]*StoredMessage{msg}, q.messages...)
	}
	return true
}

// DeadLetter moves a processing message to the given DLQ with death headers.
func (q *Queue) DeadLetter(id string, dlq *Queue, reason string) bool {
	q.mu.Lock()
	msg, ok := q.processing[id]
	if !ok {
		q.mu.Unlock()
		return false
	}
	delete(q.processing, id)
	q.mu.Unlock()

	if reason == "" {
		reason = "max retries exceeded"
	}
	headers := core.MessageHeaders{}
	for k, v := range msg.Headers {
		headers[k] = v
	}
	headers["x-original-queue"] = []byte(q.name)
	headers["x-death-reason"] = []byte(reason)
	headers["x-death-time"] = []byte(time.Now().UTC().Format(time.RFC3339))
	headers["x-delivery-attempts"] = []byte(itoa(msg.DeliveryAttempt))

	dlq.Enqueue(msg.Body, msg.Key, headers)
	return true
}

// Size returns the number of queued (not-yet-dequeued) messages.
func (q *Queue) Size() int {
	q.mu.Lock()
	defer q.mu.Unlock()
	return len(q.messages)
}

// ProcessingCount returns the number of in-flight messages.
func (q *Queue) ProcessingCount() int {
	q.mu.Lock()
	defer q.mu.Unlock()
	return len(q.processing)
}

// Clear removes all messages.
func (q *Queue) Clear() {
	q.mu.Lock()
	defer q.mu.Unlock()
	q.messages = nil
	q.processing = make(map[string]*StoredMessage)
}

// cleanup removes expired messages (caller holds the lock).
func (q *Queue) cleanup() {
	if q.maxAge <= 0 {
		return
	}
	cutoff := time.Now().Add(-q.maxAge)
	kept := q.messages[:0]
	for _, m := range q.messages {
		if m.Timestamp.After(cutoff) {
			kept = append(kept, m)
		}
	}
	q.messages = kept
}

func itoa(n int) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	if neg {
		n = -n
	}
	var buf [20]byte
	i := len(buf)
	for n > 0 {
		i--
		buf[i] = byte('0' + n%10)
		n /= 10
	}
	if neg {
		i--
		buf[i] = '-'
	}
	return string(buf[i:])
}
